### PR TITLE
Explore: Artist similarity graph

### PR DESCRIPTION
**Pull request by arsh331**
PR integrates the interactive artist similarity graph for GSoC and can be accessed under /explore/artist-similarity

Dependent on #2665

### CODE REVIEW

File: package.json
Comment (amCap1712) [inline package.json:98]: We don't use react-bootstap and bootstrap currently in the project, so these dependencies should be removed and the code should be replaced with normal html and css classes. Also, we are currently stuck at Bootstrap 3 and can't use Bootstrap 5.

File: frontend/js/src/explore/artist-similarity/artist-search/artistLookup.ts
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-search/artistLookup.ts:25]: remove alert. You can throw an error here and capture it in your components to show an error toast.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx:2]: I see multiple places where types are imported from component files. I think a better solution would be to create a new `types.d.ts` file inside `explore/artist-similarity` folder and put all the types there except the interfaces for props.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchDropdown.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-search/SearchDropdown.tsx:19]: You don't need to receive the `id` as a prop. Also, if there are multiple usages of this component you may want to change this to a class.

File: frontend/js/src/explore/artist-similarity/artist-panel/infoLookup.ts
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-panel/infoLookup.ts:104]: Remove the alert and catch it the component or otherwise to show the user a toast.

File: frontend/js/src/explore/artist-similarity/artist-panel/utils.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-panel/utils.tsx:15]: Seems unused, can be removed?

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:87]: You can remove the useRef and the useEffect below, useCallback will take of updating the function whenever getArtists changes.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:139]: `SearchDropdown` component seems simple enough that you can directly inline the code here. There are no other uses of the component AFAICS and it does not maintain any state either.

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/Data.tsx:119]: The dependency list needs to be updated.

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/Data.tsx:149]: The dependency list needs to be updated.

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/Data.tsx:163]: This should probably be separated into a separate function.

File: frontend/js/src/explore/artist-similarity/artist-panel/Panel.tsx
Comment (amCap1712) [inline frontend/js/src/explore/artist-similarity/artist-panel/Panel.tsx:81]: TODO. I will let you know when this is ready.

**Review by amCap1712 — CHANGES_REQUESTED** [review]
Hi! Thanks for the PR. It looks great but also many changes needed before it can be merged and deployed. This first set of review comments is about the more pressing issues. I also have multiple small suggestions about code quality but let's deal with it after the existing comments are resolved.

File: package.json
Comment (arsh331) [reply package.json:98]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx:22]: Ok fixed transformed the div into a button and applied required styling

File: frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx:1]: I changed the file to suit the project more, we do not require artist name, release data and type here. I made these props optional and am conditionally rendering these. A solution might be to change the existing component with this. But I am not sure about that.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchDropdown.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchDropdown.tsx:19]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-search/artistLookup.ts
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/artistLookup.ts:25]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-panel/utils.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-panel/utils.tsx:15]: I used the same file required for ReleaseCard so should I make any changes to it

File: frontend/js/src/explore/artist-similarity/artist-panel/infoLookup.ts
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-panel/infoLookup.ts:104]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:37]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:87]: If I use searchQuery as a dependency for the useCallback in getArtists it will update whenever searchQuery updates and does not work as expected. Therefore, we need to ensure getArtists never gets updated. 
So, a change suggested here, https://www.developerway.com/posts/debouncing-in-react, it says I could send the query directly to the getArtists without keeping any dependencies and it works fine this way. As opposed to, when I use a dependency and causing problems.
I have kept the previous solution as well. Let me know what will be better.

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/Data.tsx:149]: Fixed

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/Data.tsx:119]: Fixed

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/Data.tsx:131]: It seems like there is no easy way to separate the function. I have added the comment for it if you can please take a closer look.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (arsh331) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx
Comment (amCap1712) [reply frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx:1]: Makes sense, let's ponder over it later.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (amCap1712) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx]: I see, I read your code and the article and indeed your way is the correct and the most succint way to do this.

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (amCap1712) [reply frontend/js/src/explore/artist-similarity/Data.tsx]: Sure, I have pushed an updated version.

**Review by amCap1712 — COMMENTED** [review]
Hi! 

Thanks for the changes. The code looks in better shape now. We probably want to do a round of review by @MonkeyDo before merging though. I also found some issues with responsiveness while testing.

I would suggest to work on integrating BrainzPlayer meanwhile.

File: frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx
Comment (amCap1712) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchItem.tsx]: @arsh331, this feedback is still pending.

**Comment by MonkeyDo** [comment]
I took the liberty of merging master (and fixing merge conflicts) so that I can deploy this to beta.LB for testing.

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:3]: I'm not sure exactly how to use this properly, but for me currently the graph ends up cropped if I'm not at the expected screen size ratio:
image

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:61]: FYI, this centers the text vertically as one would expect (but I suggested an alternative option above)
```suggestion
        dominantBaseline="central"
```

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:31]: It would be nice to give these elements a class name so that we can target them.

For example, adding `cursor: pointer`  one these group elements would make it clearer that the graph is interactive

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:108]: Why the comparison to Infinity here, out of curiosity?
It's not the clearest way to ensure `score` is a number. For example maybe look at [Lodash's isFinite](https://lodash.com/docs/4.17.15#isFinite)

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:129]: It might be worth throttling handleResize with [lodash's `throttle` utility](https://lodash.com/docs/4.17.15#throttle)

You would create a function `throttledHandleResize` which would be passed as the event listener.
That function would only be called once every, say, 500 milliseconds.

File: frontend/js/src/explore/artist-similarity/artist-panel/Panel.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/artist-panel/Panel.tsx:45]:
```suggestion
    if (artistInfo?.topTrack?.recording_name) {
```

File: frontend/css/artist-similarity.less
Comment (MonkeyDo) [inline frontend/css/artist-similarity.less:132]: This doesn't look very nice on my screen; if there is not a lot of height everything is smooshed together and overlapping itself:
image

Please look into how to make this work better for other screen ratios (including mobile phones preferably)

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/Data.tsx:60]: We are not going to want this hardcoded.
However the server route should send an optional seed artist MBID (where we have `props=orjson.dumps(…` in the route) and support passing one as query parameters

File: frontend/js/src/explore/artist-similarity/Data.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/Data.tsx:184]: There should probably be a helpful message displayed on screen if the user hasn't yet selected an artist (cf. comment above about not having a default hardcoded)

**Comment by MonkeyDo** [comment]
Hi @arsh331 !
Please let us know if you can finish the PR, and if there is anything we can do to help.

**Comment by MonkeyDo** [comment]
Hello @arsh331 , please le me know if you don't have time to continue working on this, if that's the case I can take it over, but until you let us know I don't want to be rude and finish your project for you :)

**Comment by arsh331** [comment]
Hi @MonkeyDo, I am really sorry I cannot make time for finishing up right now. But I will be more than happy if you can to go ahead and finish it up. Also, I am excited to see the project being used in other pages. :)

File: frontend/js/src/explore/artist-similarity/ArtistSimilarity.tsx
Comment (anshg1214) [reply frontend/js/src/explore/artist-similarity/ArtistSimilarity.tsx]: Done :)

File: listenbrainz/webserver/views/explore.py
Comment (anshg1214) [inline listenbrainz/webserver/views/explore.py:66]: Here I'm fetching the top artist. Alternatively we can fetch the sitewide trending artist for this month / this week. Which option would be the best here? @MonkeyDo

File: frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/SimilarArtistsGraph.tsx:132]: Both the toolbar at the top and the graph underneath are looking a bit rough on smaller screens:
image

I wonder if the values for repulsivity, distanceMax and nodeSize should be different depending on the available screen size, or maybe the size of the nodes, or…
I tried looking for some examples or to see if there are other solutions, but didn't find anything useful.

I do notice that I get some cropped nodes also in other non-mobile screen sizes:

image

I wonder if it has to do with the size of the nodes being ignored, meaning the center of the top-most node is just about in frame but the node scaled from that center point is out of bounds

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (MonkeyDo) [inline frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:87]: Usually everywhere on the website we use the bootstrap css class `form-control` on all the input elements to homogenize the styling.
That might be at odds with the custom css for these inputs (i.e. where you define a fixed height for example)

Same goes for the other inputs in this page

File: listenbrainz/webserver/views/popularity_api.py
Comment (anshg1214) [reply listenbrainz/webserver/views/popularity_api.py:8]: Yes, this is because I created this endpoint in a separate PR. #2665

File: frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx
Comment (anshg1214) [reply frontend/js/src/explore/artist-similarity/artist-panel/ReleaseCard.tsx]: Fixed

File: frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx
Comment (anshg1214) [reply frontend/js/src/explore/artist-similarity/artist-search/SearchBox.tsx:88]: Fixed

**Review by MonkeyDo — APPROVED** [review]
~~93 94~~ 95 commits in and a lot of us working on this feature, I think we are finally ready to merge a v1 🎉 !

I'll note some initial future improvements:
* repulsivity / centering values can be improved and better fit depending on screen size
* consider replacing tinycolor2 with colorjs or chroma.js which are more powerful and would make calculating the difference between colors easier, potentially improving the color schemes
* selected artist MBID should be added to URL and update on each new artist, so that we can go directly to the right page with a URL
  * this should also be added to the brower history so we can have previous/next navigation 
  * this will be trivial once this part of the website has been refactored into a single-page app
* make the page playable
  * the top track(s) should just play when clicked instead of being a link to the album
  * the top album can be added to the BrainzPlayer queue once that PR is merged
* consider fetching an artist other than n°1 overall (The Beatles). Maybe a random one from the top X trending artists? Or the top artist(s) of the logged-in user? 
* The nodes get cropped on some screen sizes when there are a lot of nodes: 
   image

**Pull request merged**
